### DISCUSSION TITLE
Capturing all characters from a terminator's list

### DISCUSSION THREAD

**Discussion by Bobini1** [foonathan/lexy#112] [Q&A] [answered]
I've got the following bit of code:

```
struct TextTag
{
    static constexpr auto value = lexy::as_string<std::string> >> lexy::callback<std::string>(trimR);
    static constexpr auto rule = []{
        auto end =
          dsl::terminator(dsl::peek(dsl::eol));
        auto value = end.list(dsl::capture(dsl::unicode::character)); // this line is important
        return value;
    }();
};
```

It gets all characters in a line and then trims the result.
However, `dsl::capture` discards whitespace. I don't want that. `dsl::identifier` doesn't accept `dsl::peek(dsl::eol)` as the end condition so I can't use it either and I need to handle eof correctly. If I use neither of those and simply make a list of characters, nothing gets added to the sink.

I have spent like 5 hours on this already, is it really so hard to just get all characters in the sink literally?

**Comment by foonathan** [comment, marked as answer]
There are three solutions:

1. Inherit from `lexy::token_production` and leave everything unchanged. This disables whitespace skipping inside the production. https://lexy.foonathan.net/playground/?id=xe3zaajev&mode=tree

2. Use `dsl::delimited()`, which is designed for that sort of thing: https://lexy.foonathan.net/playground/?id=esnej4zjT&mode=tree

3. If you don't care about validation of Unicode code points, you can also use `dsl::capture(dsl::until(dsl::newline).or_eof())`, but note that it includes the trailing newline: https://lexy.foonathan.net/playground/?id=abE6aMMs3&mode=tree

**Reply by Bobini1** [reply]
Thank you!

**Reply by foonathan** [reply]
For reference and future readers, I've just added a third possibility.
